File: app/src/main/java/s15666/pjwstk/pamobmi/ui/bmi/BmiResultUpdater.java
package s15666.pjwstk.pamobmi.ui.bmi;

import android.view.View;
import android.widget.TextView;

import s15666.pjwstk.pamobmi.R;
import s15666.pjwstk.pamobmi.bmi.BmiViewModel;
import s15666.pjwstk.pamobmi.ui.bmi.BmiCategory.NoSuchBmiException;

class BmiResultUpdater {

    private View view;
    private BmiViewModel model;

    private TextView bmiResultField;
    private TextView bmiCategoryField;

    BmiResultUpdater(View view, BmiViewModel model, TextView bmiResultField, TextView bmiCategoryField) {
        this.view = view;
        this.model = model;
        this.bmiResultField = bmiResultField;
        this.bmiCategoryField = bmiCategoryField;

        updateResultField((double) 0);
    }

    void update(double bmi) {

        updateResultField(bmi);

        try {
            BmiCategory category = BmiCategory.getCategory(bmi);

            model.setBmiCategory(category);
            bmiCategoryField.setText(category.toString());
        } catch (NoSuchBmiException e) {
            model.setBmiCategory(null);
            bmiCategoryField.setText("");
        }
    }

    void clear() {
        updateResultField(0);
        bmiCategoryField.setText("");
    }

    private void updateResultField(double bmi) {
        bmiResultField.setText(view.getContext().getString(R.string.bmi_value, bmi));
    }
}

File: app/src/main/java/s15666/pjwstk/pamobmi/ui/bmi/BmiResultUpdater.kt
package s15666.pjwstk.pamobmi.ui.bmi

import android.view.View
import android.widget.TextView
import s15666.pjwstk.pamobmi.R
import s15666.pjwstk.pamobmi.bmi.BmiViewModel
import s15666.pjwstk.pamobmi.ui.bmi.BmiCategory.NoSuchBmiException

internal class BmiResultUpdater(
        private val view: View,
        private val model: BmiViewModel,
        private val bmiResultField: TextView,
        private val bmiCategoryField: TextView) {

    init {
        updateResultField(0.toDouble())
    }

    fun update(bmi: Double) {
        updateResultField(bmi)
        try {
            val category: BmiCategory = BmiCategory.Companion.getCategory(bmi)
            model.setBmiCategory(category)
            bmiCategoryField.text = category.toString()
        } catch (e: NoSuchBmiException) {
            bmiCategoryField.text = ""
        }
    }

    fun clear() {
        updateResultField(0.0)
        bmiCategoryField.text = ""
    }

    private fun updateResultField(bmi: Double) {
        bmiResultField.text = view.context.getString(R.string.bmi_value, bmi)
    }

}